File: chambers/src/main/java/vip/qoriginal/quantumplugin/chambers/ChamberPosition.java
package vip.qoriginal.quantumplugin.chambers;

import org.bukkit.Location;
import org.bukkit.World;

public record ChamberPosition(
        double x,
        double y,
        double z,
        float yaw,
        float pitch
) {
    public Location in(World world) {
        return new Location(world, x, y, z, yaw, pitch);
    }

    public Location relativeTo(World world, Location origin) {
        return new Location(
                world,
                origin.getX() + x,
                origin.getY() + y,
                origin.getZ() + z,
                yaw,
                pitch
        );
    }
}

File: chambers/src/main/java/vip/qoriginal/quantumplugin/chambers/data/ChamberPosition.kt
package vip.qoriginal.quantumplugin.chambers.data

import org.bukkit.Location
import org.bukkit.World

data class ChamberPosition(
    val x: Double,
    val y: Double,
    val z: Double,
    val yaw: Float,
    val pitch: Float,
) {
    fun inWorld(world: World): Location = Location(world, x, y, z, yaw, pitch)

    fun relativeTo(world: World, origin: Location): Location = Location(
	    world,
	    origin.x + x,
	    origin.y + y,
	    origin.z + z,
	    yaw,
	    pitch,
    )
}